notes/compilation-and-linkage.md

```diff
@@ -0,0 +1,684 @@
+
+Let's look at an example implementation. Here we have `point.hpp`, `point.cpp`, and `pointMain.cpp`
+
+```C++
+#define SHIFT_X = 4;
+#define SHIFT_Y = 5;
+
+#define <iostream>
+#define <string>
+
+struct point {
+    std::string name;
+    int x;
+    int y;
+};
+
+point DefaultTranslate();
+
+point Translate(std::string name, int xShift, int yShift);
+
+void printString(point p);
+```
+<div style="text-align: center">
+
+Figure 6a - `point.hpp`
+
+</div>
+
+```C++
+#include "point.hpp"
+
+point Translate(std::string name, int xShift, int yShift) {
+    if(xShift == 0 && yShift == 0) {
+        return DefaultTranslate(name);
+    } else {
+        return {name, x + xShift, y + yShift};
+    }
+}
+
+point DefaultTranslate(std::string name) {
+    return {name, x + SHIFT_X, y + SHIFT_Y};
+}
+
+void printString(point p) {
+    std::cout << Coordinate Points for << p.name << ": (" << x << ", " << y << ")" << std::endl;;
+}
+```
+<div style="text-align: center">
+
+Figure 6b - `point.cpp`
+
+</div>
+
+```C++
+#include "point.hpp"
+
+int main() {
+    point myPoint = {"myPoint", 3, 10};
+    myPoint = Translate(myPoint.name, 0, 0);
+    printString();
+}
+/* Output:
+ * Coordinate Points for myPoint: (7, 15)
+**/
+```
+<div style="text-align: center">
+
+Figure 6c - `pointMain.cpp`
+
+</div>
+A few things happen due to this "abstraction"
+
+1. Readibility - My program makes a lot more sense when I read it. My `point.hpp` tell me what the file does, and my `point.cpp` is just implementation information. Also, my `point.cpp` file only has one directive, which is just the `#include point.hpp` statement.
+2. Access - Because of the `#include point.hpp`, my `point.cpp` can access any directives inside of `point.hpp`, and so can other files now too by doing the same, like `pointMain.cpp`.
+3. Order - The ordering of function/class defintions no longer matter. As long as `#include point.hpp` goes at the top, their definitions can be seen by all entities.
+4. Memory - When the preprocessor runs, `point.cpp` and `pointMain.cpp` will have less code copied at the top of their programs.
+5. Completeness - `point.hpp` helps us make sure that we implement everything inside of it. It doesn't give an error if we don't, but there are tools out there that will tell you whether things inside of `.hpp` files have been implemented yet.
+
+However, what if included `iostream` in `pointMain.cpp`? Assuming that it doesn't have what we're about to talk about next, this will cause `iostream` to be repeated inside of `main.cpp`, which is a big problem. So, developers have created something called a header guard, WHICH YOU MUST USE!!! To use a header guard, you can employ the following to all your .hpp files
```

Comment:
I think it's good to emphasize explicitly that a main point of header guards is to guard against multiple definitions